crates/zkevm_circuits/src/bn254/ec_pairing/mod.rs

```diff
@@ -0,0 +1,556 @@
+            BN256Fq,
+            NonNativeFieldOverU16<_, bn256::Fq, 17>,
+            BN256Extension12Params,
+        > as NonNativeField<F, BN256Fq>>::conditionally_select(
+            cs,
+            state.read_words_for_round,
+            &acc,
+            &state.pairing_inner_state,
+        );
+
+        let no_pairs_left = state.precompile_call_params.num_pairs.is_zero(cs);
+        let write_result = Boolean::multi_and(cs, &[state.read_words_for_round, no_pairs_left]);
+
+        let success_as_u32 = unsafe { UInt32::from_variable_unchecked(success.get_variable()) };
+        let mut success = zero_u256;
+        success.inner[0] = success_as_u32;
+
+        let success_query = MemoryQuery {
+            timestamp: timestamp_to_use_for_write,
+            memory_page: state.precompile_call_params.output_page,
+            index: state.precompile_call_params.output_offset,
+            rw_flag: boolean_true,
+            is_ptr: boolean_false,
+            value: success,
+        };
+
+        let _ = memory_queue.push(cs, success_query, write_result);
+
+        state.precompile_call_params.output_offset = unsafe {
+            state
+                .precompile_call_params
+                .output_offset
+                .increment_unchecked(cs)
+        };
+
+        let paired = acc.sub(cs, &mut one_fq12.clone()).is_zero(cs);
+        let paired_as_u32 = unsafe { UInt32::from_variable_unchecked(paired.get_variable()) };
+        let mut paired = zero_u256;
+        paired.inner[0] = paired_as_u32;
+
+        let write_query = MemoryQuery {
+            timestamp: state.timestamp_to_use_for_write,
+            memory_page: state.precompile_call_params.output_page,
+            index: state.precompile_call_params.output_offset,
+            rw_flag: boolean_true,
+            is_ptr: boolean_false,
+            value: paired,
+        };
+
+        memory_queue.push(cs, write_query, write_result);
+
+        let input_is_empty = precompile_calls_queue.is_empty(cs);
+        let input_is_not_empty = input_is_empty.negated(cs);
+        let nothing_left = Boolean::multi_and(cs, &[write_result, input_is_empty]);
+        let process_next = Boolean::multi_and(cs, &[write_result, input_is_not_empty]);
+
+        state.read_precompile_call = process_next;
+        state.completed = Boolean::multi_or(cs, &[nothing_left, state.completed]);
+        let t = Boolean::multi_or(cs, &[state.read_precompile_call, state.completed]);
+        state.read_words_for_round = t.negated(cs);
+
+        if crate::config::CIRCUIT_VERSOBE {
+            dbg!(state.witness_hook(cs)());
+            dbg!(precompile_calls_queue.into_state().witness_hook(cs)());
+        }
+    }
+
+    if crate::config::CIRCUIT_VERSOBE {
+        dbg!(state.witness_hook(cs)());
+        dbg!(precompile_calls_queue.into_state().witness_hook(cs)());
+    }
+
+    precompile_calls_queue.enforce_consistency(cs);
+
+    state
+}
+
+pub fn ecpairing_function_entry_point<
```

Comment:
Rewrite this function. You use a bunch of code places that we don't use for pairing ( I assume you copy-paste from a hash entry point)

Reply:
We indeed reused design for `ecpairing` from hash circuits because we need to process different rounds and aggregate result, but I can not say which things are redundant on spot. Is there anything specific?

Reply:
Why do you need different rounds for pairing?